File: go/market/market_service_test.go
// SPDX-License-Identifier: MIT
package market

import (
	"context"
	"math"
	"testing"
)

func TestMarketplaceLifecycle(t *testing.T) {
	ctx := context.Background()
	m := NewInMemoryMarketService()

	var received *MarketListing
	m.OnListingReceived = func(l *MarketListing) { received = l }

	l, err := m.CreateListing(ctx, "seller1", "Bicycle", "Red mountain bike", 1500, "k3vf9z", CategoryGoods)
	if err != nil {
		t.Fatal(err)
	}
	if l.ListingID == "" {
		t.Fatal("listing has no id")
	}
	if received == nil || received.ListingID != l.ListingID {
		t.Fatal("ListingReceived not fired")
	}

	// BrowseNearby: prefix match vs a far geohash.
	near, _ := m.BrowseNearby(ctx, "k3vf9z", 2)
	if len(near) != 1 {
		t.Fatalf("expected 1 nearby, got %d", len(near))
	}
	far, _ := m.BrowseNearby(ctx, "xxxxxx", 2)
	if len(far) != 0 {
		t.Fatalf("expected 0 far, got %d", len(far))
	}

	// Search by text + category filter.
	if res, _ := m.Search(ctx, "bike", nil); len(res) != 1 {
		t.Fatalf("search 'bike' -> %d", len(res))
	}
	svc := CategoryServices
	if res, _ := m.Search(ctx, "bike", &svc); len(res) != 0 {
		t.Fatalf("search 'bike' in Services -> %d", len(res))
	}

	// Trade state machine: Initiated -> BuyerConfirmed -> Complete.
	esc, _ := m.InitiateTrade(ctx, l, "buyer1")
	if esc.State != StateInitiated {
		t.Fatalf("state %d != Initiated", esc.State)
	}
	esc, _ = m.ConfirmTrade(ctx, esc, RoleBuyer)
	if esc.State != StateBuyerConfirmed {
		t.Fatalf("state %d != BuyerConfirmed", esc.State)
	}
	esc, _ = m.ConfirmTrade(ctx, esc, RoleSeller)
	if esc.State != StateComplete {
		t.Fatalf("state %d != Complete", esc.State)
	}

	// Dispute path.
	esc2, _ := m.InitiateTrade(ctx, l, "buyer2")
	if err := m.Dispute(ctx, esc2, "item not as described"); err != nil {
		t.Fatal(err)
	}
	if esc2.State != StateDisputed {
		t.Fatalf("state %d != Disputed", esc2.State)
	}
}

func TestPoVServiceScoreAndDefection(t *testing.T) {
	p, err := NewInMemoryPoVService()
	if err != nil {
		t.Fatal(err)
	}

	tok, _ := p.IssueToken("w1", "A", TransportBle)
	if !p.VerifyToken(tok) {
		t.Fatal("issued token must verify")
	}
	if err := p.AcceptToken(tok); err != nil {
		t.Fatal(err)
	}

	sc := p.GetScore("A")
	if sc.UniqueWitnesses != 1 {
		t.Fatalf("unique witnesses %d != 1", sc.UniqueWitnesses)
	}
	if math.Abs(sc.WeightedScore-0.5) > 1e-9 { // 1/(1+1)
		t.Fatalf("weighted score %v != 0.5", sc.WeightedScore)
	}

	// Tampering invalidates the signatures.
	bad := *tok
	bad.SubjectUhid = "C"
	if p.VerifyToken(&bad) {
		t.Fatal("tampered token must not verify")
	}

	// A node cannot vouch for itself.
	self, _ := p.IssueToken("x", "x", TransportNfc)
	if p.VerifyToken(self) {
		t.Fatal("self-vouch must not verify")
	}

	// Defection penalty: A's score 0.5 -> 0.4.
	p.ReportDefection("A", "victim")
	if got := p.GetScore("A").WeightedScore; math.Abs(got-0.4) > 1e-9 {
		t.Fatalf("post-defection score %v != 0.4", got)
	}
}

File: kotlin/src/test/kotlin/aethernet/market/MarketServiceTest.kt
// SPDX-License-Identifier: MIT
//
// Behavioural tests for the in-memory aether-market services: the marketplace
// (create/browse/search + the trade-escrow state machine) and the single-node
// Proof-of-Vicinity service (issue/verify/accept/score + the defection penalty).

package aethernet.market

import org.junit.jupiter.api.Test
import kotlin.math.abs
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class MarketServiceTest {

    @Test
    fun marketplaceLifecycle() {
        val m = InMemoryMarketService()
        var received: MarketListing? = null
        m.onListingReceived = { received = it }

        val l = m.createListing("seller1", "Bicycle", "Red mountain bike", 1500.0, "k3vf9z", MarketCategory.Goods)
        assertTrue(l.listingId.isNotEmpty())
        assertEquals(l.listingId, received?.listingId)

        assertEquals(1, m.browseNearby("k3vf9z", 2).size)
        assertEquals(0, m.browseNearby("xxxxxx", 2).size)
        assertEquals(1, m.search("bike").size)
        assertEquals(0, m.search("bike", MarketCategory.Services).size)

        var e = m.initiateTrade(l, "buyer1")
        assertEquals(TradeState.Initiated, e.state)
        e = m.confirmTrade(e, TradeRole.Buyer)
        assertEquals(TradeState.BuyerConfirmed, e.state)
        e = m.confirmTrade(e, TradeRole.Seller)
        assertEquals(TradeState.Complete, e.state)

        val e2 = m.initiateTrade(l, "buyer2")
        m.dispute(e2, "item not as described")
        assertEquals(TradeState.Disputed, e2.state)
    }

    @Test
    fun povScoreAndDefection() {
        val p = InMemoryPoVService()

        val tok = p.issueToken("w1", "A", PoVTransportType.Ble)
        assertTrue(p.verifyToken(tok))
        p.acceptToken(tok)

        val sc = p.getScore("A")
        assertEquals(1, sc.uniqueWitnesses)
        assertTrue(abs(sc.weightedScore - 0.5) < 1e-9)

        // Tampering with the body invalidates the signatures.
        val bad = tok.copy(subjectUhid = "C")
        assertFalse(p.verifyToken(bad))

        // A node cannot vouch for itself.
        val self = p.issueToken("x", "x", PoVTransportType.Nfc)
        assertFalse(p.verifyToken(self))

        // Defection penalty: A's score 0.5 -> 0.4.
        p.reportDefection("A", "victim")
        assertTrue(abs(p.getScore("A").weightedScore - 0.4) < 1e-9)
    }
}
